programs/staking/src/lib.rs

```diff
@@ -1,112 +1,214 @@
-
-    let reward = pool
-        .reward_per_token(total_staked, last_time_reward_applicable)
-        .ok_or(ErrorCode::MathOverFlow)?;
-
-    pool.reward_per_token_stored = reward;
-    pool.last_update_time = last_time_reward_applicable;
-
-    if let Some(u) = user {
-        let amount = pool.user_earned_amount(u).ok_or(ErrorCode::MathOverFlow)?;
-
-        u.reward_per_token_pending = amount;
-        u.reward_per_token_complete = pool.reward_per_token_stored;
-    }
-
-    Ok(())
-}
-
 /// Single farming program
 #[program]
-pub mod staking {
+pub mod single_farming {
     use super::*;
 
     /// Initializes a new pool
     pub fn initialize_pool(
         ctx: Context<InitializePool>,
-        reward_duration: u64,
-        funding_amount: u64,
+        jup_reward_duration: u64,
+        jup_funding_amount: u64,
+        xmer_reward_duration: u64,
     ) -> Result<()> {
-        if reward_duration == 0 {
-            return Err(ErrorCode::DurationCannotBeZero.into());
+        if jup_reward_duration == 0 {
+            return Err(ErrorCode::JupDurationCannotBeZero.into());
         }
 
         let pool = &mut ctx.accounts.pool;
         // This is safe as long as the key matched the account in InitializePool context
         pool.staking_vault_nonce = *ctx.bumps.get("staking_vault").unwrap();
         pool.staking_mint = ctx.accounts.staking_mint.key();
         pool.staking_vault = ctx.accounts.staking_vault.key();
-        pool.reward_mint = ctx.accounts.reward_mint.key();
-        pool.reward_vault = ctx.accounts.reward_vault.key();
-        pool.reward_duration = reward_duration;
+
+        // update jup  info
+        pool.jup_reward_duration = jup_reward_duration;
         pool.total_staked = 0;
-        pool.last_update_time = 0;
-        pool.reward_end_timestamp = 0;
+        pool.jup_last_update_time = 0;
+        pool.jup_reward_end_timestamp = 0;
         pool.admin = ctx.accounts.admin.key();
-        pool.reward_rate =
-            rate_by_funding(funding_amount, reward_duration).ok_or(ErrorCode::MathOverFlow)?;
-        pool.reward_per_token_stored = 0;
+        pool.jup_reward_rate = rate_by_funding(jup_funding_amount, jup_reward_duration)
+            .ok_or(ErrorCode::MathOverFlow)?;
+        pool.jup_reward_per_token_stored = 0;
+
+        // update xmer info
+        pool.xmer_reward_duration = xmer_reward_duration;
+        pool.xmer_reward_mint = ctx.accounts.xmer_reward_mint.key();
+        pool.xmer_reward_vault = ctx.accounts.xmer_reward_vault.key();
+        pool.xmer_last_update_time = 0;
+        pool.xmer_reward_end_timestamp = 0;
+        pool.xmer_reward_per_token_stored = 0;
         Ok(())
     }
 
-    /// Admin activates farming
-    pub fn activate_farming<'a, 'b, 'c, 'info>(
-        ctx: Context<'a, 'b, 'c, 'info, ActivateFarming<'info>>,
+    /// Admin activates jup farming
+    pub fn activate_jup_farming<'a, 'b, 'c, 'info>(
```

Comment:
I think we need to prevent this function to be called twice. As it will extend the `jup_reward_end_timestamp` and doesn't align with the reward rate.

Reply:
correct, lets me fix this

Reply:
Fixed